presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughProject.java

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.optimizations.SymbolMapper;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Optional;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - Project (non-identity)
+ *       - Source other than Filter(TableScan) / TableScan
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project
+ *    - TopN
+ *       - Source
+ * </pre>
+ */
+public final class PushTopNThroughProject
+        implements Rule<TopNNode>
+{
+    private static final Capture<ProjectNode> PROJECT_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN()
+                    .with(source().matching(
+                            project()
+                                    .matching(projectNode -> !projectNode.isIdentity())
+                                    .capturedAs(PROJECT_CHILD)
+                                    .with(source().matching(node -> !(node instanceof TableScanNode)))));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode parent, Captures captures, Context context)
+    {
+        ProjectNode projectNode = captures.get(PROJECT_CHILD);
+
+        PlanNode projectSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst().orElseThrow(IllegalStateException::new);
+        if (projectSource instanceof FilterNode) {
+            PlanNode filterSource = context.getLookup().resolveGroup(((FilterNode) projectSource).getSource()).findFirst().orElseThrow(IllegalStateException::new);
```

Comment:
Please format this as:
```
PlanNode filterSource = context.getLookup().resolveGroup(((FilterNode) projectSource).getSource())
    .findFirst()
    .orElseThrow(IllegalStateException::new);
```

Same above.

Also instead of `findFirst` you should use `com.google.common.collect.MoreCollectors#toOptional`.

Reply:
`Lookup#resolveGroup` followed by `.findFirst()` (or `toOptional`) is _only pretending_ we support non-single element groups here. 
It would be way better to use the (deprecated) `io.prestosql.sql.planner.iterative.Lookup#resolve` directly instead of _pretending_ we do something better.